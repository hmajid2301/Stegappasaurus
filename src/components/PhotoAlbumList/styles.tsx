import {Dimensions, StyleSheet} from 'react-native';

const pageWidth = Dimensions.get('window').width;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginRight: 2,
  },

  photoButton: {
    flex: 1,
  },

  photos: {
    height: pageWidth / 3,
    marginLeft: 2,
    marginTop: 2,
  },
});

export default styles;
