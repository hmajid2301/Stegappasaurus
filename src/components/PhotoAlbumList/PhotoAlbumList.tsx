import * as MediaLibrary from "expo-media-library";
import * as Permissions from "expo-permissions";
import React from "react";
import { FlatList, Image, TouchableOpacity, View } from "react-native";

import Snackbar from "~/components/Snackbar";
import styles from "./styles";

interface IProps {
  onPhotoPress: (uri: string) => any;
}

interface IState {
  lastPhoto: string;
  photos: IPhoto[];
  refreshing: boolean;
}

interface IPhoto {
  uri: string;
}

export default class PhotoAlbumList extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      lastPhoto: "",
      photos: [],
      refreshing: false
    };
  }

  public render() {
    return (
      <View>
        <FlatList
          data={this.padData(this.state.photos)}
          keyExtractor={(item, index) => item.uri + index}
          numColumns={3}
          onEndReached={this.morePhotosFromCameraRoll}
          onRefresh={this.handleRefresh}
          renderItem={this.renderPhotosFromCameraRoll}
          refreshing={this.state.refreshing}
        />
      </View>
    );
  }

  public async componentDidMount() {
    setTimeout(async () => {
      await this.getPhotosFromCameraRoll();
    }, 1000);
  }

  private padData(data: IPhoto[]) {
    const itemsPerColumns = 3;
    const itemsLeftOver = data.length % itemsPerColumns;
    const elementsToAdd = itemsLeftOver === 0 ? 0 : 3 - itemsLeftOver;

    for (let i = 0; i < elementsToAdd; i += 1) {
      data.push({ uri: "" });
    }

    return data;
  }

  private handleRefresh = async () => {
    this.setState({ refreshing: true });
    await this.getPhotosFromCameraRoll();
  };

  private morePhotosFromCameraRoll = async () => {
    const { assets, endCursor } = await MediaLibrary.getAssetsAsync({
      after: this.state.lastPhoto,
      first: 9,
      sortBy: ["creationTime"]
    });

    this.setState({
      lastPhoto: endCursor,
      photos: [...this.state.photos, ...assets]
    });
  };

  private renderPhotosFromCameraRoll = ({ item }: { item: IPhoto }) => {
    if (item.uri === "") {
      return <View />;
    }

    return (
      <TouchableOpacity
        onPress={() => this.props.onPhotoPress(item.uri)}
        style={styles.photoButton}
      >
        <Image source={{ uri: item.uri }} style={styles.photos} />
      </TouchableOpacity>
    );
  };

  private getPhotosFromCameraRoll = async () => {
    const asked = await Permissions.getAsync(
      Permissions.CAMERA,
      Permissions.CAMERA_ROLL,
      Permissions.NOTIFICATIONS
    );

    if (asked.status !== "denied") {
      const { status } = await Permissions.askAsync(
        Permissions.CAMERA,
        Permissions.CAMERA_ROLL,
        Permissions.NOTIFICATIONS
      );
      if (status === "granted") {
        const { assets, endCursor } = await MediaLibrary.getAssetsAsync({
          first: 15,
          sortBy: [MediaLibrary.SortBy.creationTime]
        });

        this.setState({
          lastPhoto: endCursor,
          photos: assets,
          refreshing: false
        });
      } else {
        Snackbar.show({
          text: "Grant permissions to access camera roll, to view photos here."
        });
      }
    }
  };
}
