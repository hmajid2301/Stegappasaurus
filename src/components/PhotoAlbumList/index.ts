import PhotoAlbumList from "./PhotoAlbumList";

export default PhotoAlbumList;
