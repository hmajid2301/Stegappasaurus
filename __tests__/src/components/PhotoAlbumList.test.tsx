import { shallow, ShallowWrapper } from "enzyme";
import React from "react";

import PhotoAlbumList from "~/components/PhotoAlbumList";

describe("PhotoAlbumList: Match snapshots", () => {
  test("1", () => {
    const component = shallow(
      <PhotoAlbumList onPhotoPress={() => jest.fn()} />
    );
    expect(component).toMatchSnapshot();
  });

  test("2", () => {
    const data = {
      assets: [
        {
          mediaType: "photo",
          modificationTime: 1561754347000,
          creationTime: 1561754346927,
          width: 480,
          uri:
            "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163906_2.jpg",
          id: "40",
          filename: "IMG_20190628_163906_2.jpg",
          albumId: "-1739773001",
          height: 480,
          duration: 0
        }
      ],
      endCursor: "24"
    };

    const component = shallow(
      <PhotoAlbumList onPhotoPress={() => jest.fn()} />
    );
    expect(component).toMatchSnapshot();
  });
});

describe("PhotoAlbumList: Functions", () => {
  let instance: React.Component<{}, {}, any>;
  let component: ShallowWrapper<
    any,
    Readonly<{}>,
    React.Component<{}, {}, any>
  >;

  beforeAll(() => {
    component = shallow(<PhotoAlbumList onPhotoPress={() => jest.fn()} />);
    instance = component.instance();
  });

  let testData: any = [
    {
      data: [],
      expected: []
    },
    {
      data: [{ uri: "a.png" }],
      expected: [{ uri: "a.png" }, { uri: "" }, { uri: "" }]
    },
    {
      data: [{ uri: "a.png" }, { uri: "b.png" }],
      expected: [{ uri: "a.png" }, { uri: "b.png" }, { uri: "" }]
    },
    {
      data: [{ uri: "a.png" }, { uri: "b.png" }, { uri: "c.png" }],
      expected: [{ uri: "a.png" }, { uri: "b.png" }, { uri: "c.png" }]
    }
  ];
  test.each(testData)("padData", ({ data, expected }) => {
    const paddedData = (instance as any).padData(data);
    expect(paddedData).toEqual(expected);
  });

  testData = {
    assets: [
      {
        mediaType: "photo",
        modificationTime: 1561754349000,
        creationTime: 1561754349228,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163909.jpg",
        id: "47",
        filename: "IMG_20190628_163909.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754349000,
        creationTime: 1561754348941,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163908_2.jpg",
        id: "46",
        filename: "IMG_20190628_163908_2.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754348000,
        creationTime: 1561754348528,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163908_1.jpg",
        id: "45",
        filename: "IMG_20190628_163908_1.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754348000,
        creationTime: 1561754348176,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163908.jpg",
        id: "44",
        filename: "IMG_20190628_163908.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754348000,
        creationTime: 1561754347902,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163907_2.jpg",
        id: "43",
        filename: "IMG_20190628_163907_2.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754347000,
        creationTime: 1561754347517,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163907_1.jpg",
        id: "42",
        filename: "IMG_20190628_163907_1.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754347000,
        creationTime: 1561754347328,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163907.jpg",
        id: "41",
        filename: "IMG_20190628_163907.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        mediaType: "photo",
        modificationTime: 1561754347000,
        creationTime: 1561754346927,
        width: 480,
        uri: "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163906_2.jpg",
        id: "40",
        filename: "IMG_20190628_163906_2.jpg",
        albumId: "-1739773001",
        height: 480,
        duration: 0
      },
      {
        uri: ""
      }
    ],
    endCursor: "24"
  };

  test("morePhotosFromCameraRoll", async () => {
    component.setState({ photos: [] });
    await (instance as any).morePhotosFromCameraRoll();
    expect(component.state("lastPhoto")).toEqual("24");
    expect(component.state("photos")).toEqual(testData.assets);
  });

  test("renderPhotosFromCameraRoll", () => {
    component.setState({ photos: [] });
    const snap = (instance as any).renderPhotosFromCameraRoll({
      item: testData.assets[0]
    });
    const snap2 = (instance as any).renderPhotosFromCameraRoll({
      item: testData.assets[8]
    });
    expect(snap).toMatchSnapshot();
    expect(snap2).toMatchSnapshot();
  });

  test("getPhotosListFromAlbum: granted", async () => {
    component.setState({ photos: [], refreshing: true });
    await (instance as any).getPhotosFromCameraRoll();
    expect(component.state("lastPhoto")).toEqual("24");
    expect(component.state("photos")).toEqual(testData.assets);
    expect(component.state("refreshing")).toEqual(false);
  });
});
