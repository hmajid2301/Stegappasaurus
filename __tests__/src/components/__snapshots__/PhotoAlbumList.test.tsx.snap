// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoAlbumList: Functions renderPhotosFromCameraRoll 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "file:///storage/emulated/0/DCIM/Camera/IMG_20190628_163909.jpg",
      }
    }
    style={
      Object {
        "height": 250,
        "marginLeft": 2,
        "marginTop": 2,
      }
    }
  />
</TouchableOpacity>
`;

exports[`PhotoAlbumList: Functions renderPhotosFromCameraRoll 2`] = `<View />`;

exports[`PhotoAlbumList: Match snapshots 1 1`] = `
<View>
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onRefresh={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`PhotoAlbumList: Match snapshots 2 1`] = `
<View>
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onRefresh={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
